### ISSUE TITLE
When will this mod be complete?

### ISSUE THREAD

**Issue by asiekierka** [diesieben07/CameraCraft3#1]
I would like to offer my coding assistance. I miss this mod greatly from 1.4.7.

**Comment by diesieben07** [comment]
Soon™. If you wish to help, get in touch. I am currently working on updating it, which has been held back until now because of a dependency.

**Comment by NimmerNeko** [comment]
Oh man how i love the term soon :D i cant at all unfortunatly but would help if i could. is it v 1.7 or 1.8 you are going for

**Comment by diesieben07** [comment]
First 1.7.10 then 1.8 soon²™.

**Comment by NimmerNeko** [comment]
Gr8 M8. I know you modders hate ETA's but can i get one anyways? and if you need beta testers (or alpha testers) i can help. I know enough about modding to say what is broken when i find a mistake.

**Comment by NimmerNeko** [comment]
I can 3d model a bit too

**Comment by diesieben07** [comment]
The ETA is Soon™. Sorry.

**Comment by NimmerNeko** [comment]
-.-

**Comment by rihlsul** [comment]
"You keep trying to ~~pick a fight~~ _get an ETA_, but I'm still just happy ~~you're~~ _this project's_ alive."

**Comment by Docithe** [comment]
Been looking forward to this for a while! I'm happy to see that it's still alive. :)

I don't have a lot of Minecraft modding experience, but I am a professional java developer and avid MC player. If you are still interested in any help, I'd be happy to do what I can in my spare time.

**Comment by HyperMan19680** [comment]
It's been a while, any news?

**Comment by diesieben07** [comment]
I don't know what to tell you. I know it sucks to keep repeating the same answer and no progress. But... I struggle with keeping a project going and finishing it in decent time, so without turning this into a psychology session I cannot give you an ETA.

**Comment by HyperMan19680** [comment]
Jeez, I wish I could help. But I have no Java knowledge. All I can suggest is to get help. I think iChun could help as I saw some potential in their Doors mod.

**Comment by diesieben07** [comment]
The biggest problem is that the core principle of placing down cameras completely goes against Minecraft's design. If the player walks away from the camera, I have no way of rending the viewpoint of the camera. And even less so, if all players log off the server.
That said, almost everything else is done :D

**Comment by HyperMan19680** [comment]
I think this is where coremods and chunk-loaders can come into play. Coremods I feel can add more to Minecraft, like your viewpoint idea. And you can have the cameras chunkload so that the chunk they are in stays loaded.

**Comment by diesieben07** [comment]
I know a _lot_ about coremods, trust me. And yes, you can chunkload, but chunk-loading is completely server-side. The server has no idea how the world _looks_, it cannot render the view of a camera into an image. Only the client's engine can do that (the server may not even have a graphics card!). So no matter how many hoops and loops you jump though, you cannot get around the fact that if you want a view of the world (on screen or into an image) you _need_ a Minecraft client rendering the world.
I have a very buggy, very hacky system at the moment which allows the client to load chunks outside the players view (and even in other dimensions), but this is a horrible coremod which is most likely going to explode into a million pieces if it comes into contact with other mods. And it is not the most performant thing either, since the client has to keep this stuff loaded and simulated on top of the normal rendering.

**Comment by HyperMan19680** [comment]
Well we don't have to have them doing a moving image all the time. Ever hear of time-lapse?

**Comment by diesieben07** [comment]
I have investigated some more and I have found some (hopefully) viable solutions. There is the possibility of creating an OpenGL context (OpenGL is the API Minecraft uses to talk to the graphics card) without an actual graphics card, using OSMesa. I will have to investigate on how to integrated this into the Minecraft client, but it is a possibility.

**Comment by rihlsul** [comment]
Wait, are you talking about security / CC cameras?  What I've been pining for since 1.4.7 was just the:
- Camera item that can capture an image
- Printer Item that can print various size picture frame objects that can display those images.

Is that equally unfeasible?

**Comment by diesieben07** [comment]
No, what I'm talking about is cameras on tripods being activated by redstone.
What you described is already implemented.

**Comment by HyperMan19680** [comment]
Diesie, I can imagine people are patient to a point. Maybe we should start with some alpha releases and improve on them when breakthroughs come forth?

**Comment by diesieben07** [comment]
If there was usable code, there would be an alpha release ;)

**Comment by Docithe** [comment]
I'm still happy to see that this project is alive. I have been very impressed with the mod so far, and I look forward to seeing what you're doing with it! :)

**Comment by diesieben07** [comment]
So, as an update, here is what I have decided on:
For cameras that are not held by a player, there will be a config option to control their behavior:
1. Only render images for them by the player that placed them, if they are not online or not within radius of the camera, the camera will not take pictures. This will result in "world holes" if you get further away and the camera can see into the part of the world that is no longer loaded.
2. Like 1, but will choose any player that is near the camera to render the images (this will not change how those players play the game, but it might decrease their framerate, depending on how often an image is taken by the camera).
3. You can connect a dedicated "image render client" to the server, which is just a normal Minecraft client running a special version of CameraCraft. This client will be used solely to render the images needed for placed cameras.
4. Like 3, but run the "image render client" as a silent background process on the same machine as the server, using "magic pixie dust" to render stuff without a GPU (see above). This will probably be added in a later update.

**Comment by HyperMan19680** [comment]
Diesieben, are you aware of a mod being made for 1.10.2 called ShootIt?

https://minecraft.curseforge.com/projects/shootit

**Comment by diesieben07** [comment]
No, I am not.

**Comment by HyperMan19680** [comment]
That was a quick response. Maybe they can get help from ya to make their mod like yours.

**Comment by diesieben07** [comment]
Not really, I still plan on making CameraCraft again :P

**Comment by HyperMan19680** [comment]
That is good! Any news?

**Comment by diesieben07** [comment]
Nothing official, but... I am working on it and I have great plans™.

**Comment by rihlsul** [comment]
All the great plans are really good, but even just a working camera/printer system that was the core would be so awesome.  I loved that so much, so much so that I check on this thread every few months.  :D  Let us know if you need/want any help.

**Comment by Matthysse** [comment]
Is someone alive? I loved too much this mod! Please 1.7.10 release? Or is this possible to compile the source code (usable?)?

**Comment by diesieben07** [comment]
I am alive, but I have somewhat abandoned this mod. 
There is no simple way to make cameras work properly on servers without huge efforts, which would break any kind of mod-compatibility.

If someone wants to help find solutions to the following problems, I would be happy to update this mod:

- If you place down a camera and trigger it with redstone, how do we render the image? Currently this requires a player to be around, but that is not necessarily the case. What if _nobody_ is online?
- Photos use the texture pack of the player taking the photo. If a camera is placed down, this is counter-intuitive, since different players might be nearby.
- Taking pictures requires that area to be loaded on the client. If the player is not nearby, chunks have to be transferred that would normally not be loaded on the client. This is not practical, since there is no good way to do it with how Minecraft handles chunk loading.

**Comment by Docithe** [comment]
All those issues are related to placing cameras. What if you stick to the handheld cameras for now?

**Comment by diesieben07** [comment]
The texture pack issue still applies, photos would be rendered with different texture packs.
But yes, that would solve that issue. But handheld cameras are boring 😄 

I am on holiday starting next week, I'll see if I am in the mood to do some coding 😉

**Comment by rihlsul** [comment]
I would personally still just love to have the handhelds.  Way back when, I used these to do a lot of exploring with lots of mini-bases and teleporters - having the photos up near the teleporters was great instead of signs.  Same thing with being able to make art galleries in a house.

A few of the games I had this in, I also would replace the image on my local machine with generated graphics, allowing me to make adventure packs with custom instructional how-to images.  Super helpful.

Still super good memories - I would just wish the placeable/security style cameras could be an addon/child mod so that this one, which is sounds like is more do-able, could carry on.

**Comment by diesieben07** [comment]
Update: There will be a 1.12.2 release, hopefully soon. It will not have placeable cameras, because those suck to implement. If someone has a solution and/or wants to help, please step up (coding happens in Kotlin).

**Comment by Matthysse** [comment]
No 1.7.10 release? I know it's hard to downgrade, but i have a 1.7.10 RP server, and some of the best 1.7.10 mods aren't updated

**Comment by asiekierka** [comment]
@Matthysse Sorry for dropping in like this - as a 1.4.7 CameraCraft user myself I've been watching this issue for quite some time - but could you name said mods? For statistical/curiosity purposes, really.

**Comment by Matthysse** [comment]
@asiekierka Jammy Furnitures (better than mrCraysFish), AtmosMobs, old mAtmos sounds <3, Extended decorations (flower pots and better carpets) The lion king mod, old better shaders (old zeus shaders were better) zeplin mod. Omg so good mods...

**Comment by diesieben07** [comment]
No, I am not making a release for 1.7.10.

**Comment by Matthysse** [comment]
Are you still developping this mod?

**Comment by Quatroctus** [comment]
@diesieben07 a fix for basically all of this is pretty simple. Who said you had to render the image when the picture was taken? Why not just store the chunks "within the picture" as raw data and then when the client wants to access them recreate the scene and render it then and save the image in a file so it won't have to be rendered again. This solves the problems above but may cause more problems, that I haven't thought about yet.

**Comment by diesieben07** [comment]
In theory that might work, yes. But I have not looked at how feasible this idea is.
I am terrible with anything rendering related, which is not really great for developing a mod whose key feature requires some rendering stuff. I have no idea how hard it is to basically spin up a completely separate rendering pipeline to render this "fake world".
Let alone mods which store stuff in static variables or things like `Minecraft#player`, which would not make any sense in the context of the fake world.

Anyone who wants to help out with rendering,  please get in touch.

**Comment by Matthysse** [comment]
@diesieben07 when can we expect a release without pleceable cameras?
